src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTask.java

```diff
@@ -316,6 +334,35 @@ private boolean isReadyToBeDeleted(Envelope envelope) {
         ).contains(envelope.getStatus());
     }
 
+    private void handleConstraintViolation(
+        String containerName,
+        String zipFilename,
+        String leaseId,
+        ConstraintViolationException exception
+    ) {
+        log.warn(
+            "Rejected file {} from container {} - DB constraint violation {}",
+            zipFilename,
+            containerName,
+            exception.getConstraintName(),
+            exception
+        );
+
+        if (exception.getConstraintName().equals("scannable_item_dcn")) {
+            handleInvalidFileError(
+                Event.FILE_VALIDATION_FAILURE,
+                containerName,
+                zipFilename,
+                leaseId,
+                new DuplicateDocumentControlNumbersInEnvelopeException(
+                    "Received envelope with 'document_control_number' already present in the system"
+                )
```

Comment:
in all other places an exception that has been thrown is sent to this method, while here it is artificially created, why?

Reply:
Oh, I missed that. Yes, that's weird. Maybe we could have an override to handle this kind of situation. Plus the exception's name suggests the envelope contains duplicate control numbers within itself.

Reply:
Here's some idea (not well thought through). What if we catch `ConstraintViolationException` at a lower level, closer to saving the envelope and throwing `DuplicateDocumentControlNumberException` from there? Since it implements `InvalidEnvelopeException`, no special handling would be required.

Reply:
Oh is it? It can be again

Since extra layer before saving to db is used, can move as you suggest. Sounds good. Reduces main task tampering as well 👍